opacclient/opacapp/src/main/java/de/geeksfactory/opacclient/reminder/ReminderHelper.java

```diff
@@ -0,0 +1,133 @@
+package de.geeksfactory.opacclient.reminder;
+
+import org.joda.time.LocalDate;
+import org.joda.time.format.DateTimeFormat;
+
+import android.app.AlarmManager;
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+import android.util.Log;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import de.geeksfactory.opacclient.BuildConfig;
+import de.geeksfactory.opacclient.OpacClient;
+import de.geeksfactory.opacclient.objects.LentItem;
+import de.geeksfactory.opacclient.storage.AccountDataSource;
+
+public class ReminderHelper {
+    private OpacClient app;
+    private SharedPreferences sp;
+
+    public ReminderHelper(OpacClient app) {
+        this.app = app;
+        sp = PreferenceManager.getDefaultSharedPreferences(app);
+    }
+
+    /**
+     * Save alarms for expiring media to the DB and schedule them using {@link
+     * android.app.AlarmManager}
+     */
+    public void generateAlarms() {
+        int warning = Integer.parseInt(sp.getString("notification_warning", "3"));
+
+        AccountDataSource data = new AccountDataSource(app);
+        data.open();
+        List<LentItem> items = data.getAllLentItems();
+
+        // Sort lent items by deadline
+        Map<LocalDate, List<Long>> arrangedIds = new HashMap<>();
+        for (LentItem item : items) {
+            LocalDate deadline = item.getDeadline();
+            if (!arrangedIds.containsKey(deadline)) {
+                arrangedIds.put(deadline, new ArrayList<Long>());
+            }
+            arrangedIds.get(deadline).add(item.getDbId());
+        }
+
+        // Remove alarms with no corresponding media
+        for (Alarm alarm : data.getAllAlarms()) {
+            if (!arrangedIds.containsKey(alarm.deadline)) {
+                data.removeAlarm(alarm);
+            }
+        }
+
+        // Find and add/update corresponding alarms for current lent media
+        for (Map.Entry<LocalDate, List<Long>> entry : arrangedIds.entrySet()) {
+            LocalDate deadline = entry.getKey();
+            long[] media = toArray(entry.getValue());
```

Comment:
It is not a real problem, but since were doing code-review and since we are perfectionists: I don't really like the number of ``List<Long>``/``long[]``/comma-separated string conversions.

Storing comma-separated strings in a database is a [really bad](https://en.wikipedia.org/wiki/Database_normalization) design pattern in general, but **I'm absolutely okay with it here** since the alarm database works more like a cache than like a relational database. **However**, on the Java side, wouldn't be some form of ``Set`` (e.g. ``HashSet``) the appropriate representation instead of mixing arrays and lists?

So I'd suggest keeping the database as it is but use Sets throughout this file and inside the Alarm object.